keps/sig-auth/3325-self-user-attributes-review-api/README.md

```diff
@@ -0,0 +1,299 @@
+It may be obscure which user attributes the user eventually gets after all that magic happened before authentication.
+The motivation for this KEP is to reduce obscurity and help users with debugging the authentication stack.
+
+### Goals
+
+- Add the API endpoint to get user attributes
+
+### Non-Goals
+
+- Add a corresponding kubectl command
+
+## Proposal
+
+Add a new API endpoint to the `authentication` group - `SelfUserAttributesReview`.
+The user will hip the endpoint after authentication happens, so all attributes will be available to return.
+
+## Design Details
+
+This design is inspired by the `*AccessReview` and `TokenReview` APIs.
+The endpoint has no input parameters or a `spec` field because only the authentication result is required.
+
+The structure for building a request:
+```go
+type SelfUserAttributesReview struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status is filled in by the server with the user attributes.
+	Status SelfUserAttributesReviewStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+```
+
+On receiving a request, the Kubernetes API server fills the status with the user attributes and returns it to the user.
+
+```go
+type SelfUserAttributesReviewStatus struct {
+	// User attributes of the current user.
+	// +optional
+	UserInfo UserInfo `json:"userInfo,omitempty" protobuf:"bytes,1,opt,name=userInfo"`
+}
+
+type UserInfo struct {
+	Name   string              `json:"name" protobuf:"bytes,1,opt,name=name"`
+	UID    string              `json:"uid" protobuf:"bytes,2,opt,name=uid"`
+	Groups []string            `json:"groups" protobuf:"bytes,1,opt,name=groups"`
+	Extra  map[string][]string `json:"extra" protobuf:"bytes,1,opt,name=extra"`
+}
+```
+
+User attributes are known at the moment of accessing the rest API endpoint and can be extracted from the request context.
+
+### Test Plan
+
+Unit tests covering:
+
+1. Request returns all user attributes
+2. Request returns some user attributes
+3. Request with a status returns overridden fields
+
+Integration test covering:
+
+1. Successful authentication through a simple authenticator, e.g., token or certificate authenticator
+2. Successful authentication through a complicated authenticator, e.g., webhook or authentication proxy authenticator
+3. Failed authentication
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit and integration tests completed and enabled
+
+#### Beta
+
+- Gather feedback from users
+
+#### GA
+
+- Corresponding kubectl command implemented
+
+NOTE: Should not be a part pf [conformance tests](https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md).
```

Comment:
Calling out to other reviewers.  I'm ok with this position.

also typo: s/pf/of

Reply:
Good point. I extended the note.